### DISCUSSION TITLE
Why do the linebreaks in multiline @-strings use that of the .cs code file instead of `Environment.NewLine` or a constant one?

### DISCUSSION THREAD

**Discussion by oxygen-dioxide** [dotnet/csharplang#8237] [Q&A]
```C#
using System.IO;

string a = @"
";
Console.WriteLine(a.Length);
```
If I save the above code in CRLF linebreak, it will output `2`. However, if I save it in LF linebreak, it will output `1`

This makes multiline string **TOTALLY USELESS** if I want to upload my code onto GitHub, because another user (or GitHub Actions) with a different git config may get wrong result.

**Comment by oxygen-dioxide** [comment]
This is the equivalent python code:
```py
a="""
"""
print(len(a))
```
It always outputs `1` no matter which linebreak style the file uses

**Reply by HaloFour** [reply]
Raw strings contain the raw text included in the source file.  It is not reinterpretted based on the platform on which it is run.

**Comment by colejohnson66** [comment]
If you want/need a definite/consistent line ending for other developers, you can enforce it with an editor config.

**Comment by CyrusNajmabadi** [comment]
> ```C#
> using System.IO;
> 
> string a = @"
> ";
> Console.WriteLine(a.Length);
> ```
> If I save the above code in CRLF linebreak, it will output `2`. However, if I save it in LF linebreak, it will output `1`
> 
> This makes multiline string **TOTALLY USELESS** if I want to upload my code onto GitHub, because another user (or GitHub Actions) with a different git config may get wrong result. 

Your tools should not be making changes to your source files. Source files contain *content*.  So changing something like line endings literally changes the meaning of code.  

We don't use Env.NewLine for the same reason.  The meaning of your code would change on different computers. 

If you don't want your code meaning to change, don't have your tools change the content.  For something like git/GitHub there are options for this.

**Comment by CyrusNajmabadi** [comment]
Git and GitHub came out in 2005 and 2008. Many years after c# was already out.   No clue why they may have picked an option that changes the content of source files. Fortunately, there's an easy fix for that. Just set the right options in your .gitattributes file so that git doesn't munge anything inappropriately. 

See https://docs.github.com/en/get-started/getting-started-with-git/configuring-git-to-handle-line-endings for more details.

**Reply by huoyaoyuan** [reply]
It's not limited to source files.
I had a torrent file for testing in my repo. Git detects it as text file and replaces `0F0A` sequence with `0A` on Linux, causing unit test failing on Linux. This gave me a strong impression.

**Comment by jnm2** [comment]
What's the alternative? Having the length depend on whether the compiler happened to be running on Windows? Or making these string literals non-constant, rebuilding the string at runtime to insert Environment.NewLines each time the code runs?
